File: typescript/src/core/modules/auth/functions/check-teams.ts
// Import errors
import { ClientError } from "../../../error";

// Import helpers
import { getInfoFromClaims } from "../helpers/get-info-from-claims";

// Import types
import type { Pipeline } from "../../../context/pipeline";
import type { TRuntimeContext } from "../../../context/runtime-context";

/**
 * Tạo một Step Executor mới để kiểm tra team người dùng.
 *
 * @param pipeline - pipeline.
 * @param allowedTeams - các vai trò được phép thực hiện.
 *
 * @returns
 */
export function createTeamsCheckStepExecutor(
  pipeline: Pipeline<any>,
  allowedTeams: Array<string>,
) {
  return async function (ctx: TRuntimeContext) {
    if (allowedTeams[0] === "*") {
      return;
    }

    const claims = await ctx.getTempData("claims");
    const user = getInfoFromClaims(claims);

    if (!allowedTeams.find((team) => user.team === team)) {
      pipeline.stop(ctx);

      const err = new ClientError(
        "You don't have permission to do this action",
      );
      err.asHTTPError("Forbidden");
      err.addErrorDetail({
        source: "teamCheck",
        desc: `Team of user is not allowed: ${user.team}`,
      });
      return ctx.sendJson(err);
    }
  };
}

File: python/src/core/modules/auth/functions/check_teams.py
from typing import Any, Callable, List

from core.error.ClientError import ClientError
from core.context.pipeline.main import Pipeline
from core.context.runtime_context.RuntimeContext import RuntimeContext
from helpers.get_info_from_claims import get_info_from_claims


def create_teams_check_step_executor(
    pipeline: Pipeline, allowed_teams: List[str]
) -> Callable[[RuntimeContext], Any]:
    """
    Tạo một Step Executor mới để kiểm tra team người dùng.

    Args:
        pipeline: pipeline.
        allowed_teams: các vai trò được phép thực hiện.
    Returns:
        hàm step executor.
    """

    def step_executor(ctx: RuntimeContext):
        if allowed_teams and allowed_teams[0] == "*":
            return

        claims = ctx.get_temp_data("claims")
        user = get_info_from_claims(claims)

        if user.get("team") not in allowed_teams:
            pipeline.stop(ctx)

            err = ClientError("You don't have permission to do this action")
            err.as_http_error("Forbidden")
            err.add_error_detail(
                {
                    "source": "teamCheck",
                    "desc": f"Team of user is not allowed: {user.get('team')}",
                }
            )
            return ctx.send_json(err)

    return step_executor
